src/coreclr/jit/morph.cpp

```diff
@@ -481,6 +486,35 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
 #endif // TARGET_AMD64
 
 #ifdef TARGET_X86
+#ifdef FEATURE_HW_INTRINSICS
+    else if (varTypeIsLong(srcType) && varTypeIsFloating(dstType) && canUseEvexEncoding())
+    {
+        // We can avoid helper calls by using SIMD conversion instructions. The result needs to end up
+        // in a SIMD/floating register anyway.
+        NamedIntrinsic intrinsicId      = NI_Illegal;
+        CorInfoType    baseFloatingType = (dstType == TYP_FLOAT) ? CORINFO_TYPE_FLOAT : CORINFO_TYPE_DOUBLE;
+        CorInfoType    baseIntegralType = tree->IsUnsigned() ? CORINFO_TYPE_ULONG : CORINFO_TYPE_LONG;
+
+        if (compOpportunisticallyDependsOn(InstructionSet_AVX512DQ_VL))
+        {
+            intrinsicId = (dstType == TYP_FLOAT) ? NI_AVX512DQ_VL_ConvertToVector128Single
+                                                 : NI_AVX512DQ_VL_ConvertToVector128Double;
+        }
+        else
+        {
+            assert(compIsaSupportedDebugOnly(InstructionSet_AVX10v1));
+            intrinsicId =
+                (dstType == TYP_FLOAT) ? NI_AVX10v1_ConvertToVector128Single : NI_AVX10v1_ConvertToVector128Double;
+        }
+
+        GenTree* createScalar = gtNewSimdCreateScalarNode(TYP_SIMD16, oper, baseIntegralType, 16);
+        GenTree* convert      = gtNewSimdHWIntrinsicNode(TYP_SIMD16, createScalar, intrinsicId, baseIntegralType, 16);
+        GenTree* toScalar     = gtNewSimdToScalarNode(dstType, convert, baseFloatingType, 16);
```

Comment:
I believe we'd normally do this transform in lowering, but that would be complicated because of long decomposition, which happens just before. This was the simplest approach I could think of. However, [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=994741&view=ms.vss-build-web.run-extensions-tab) show that this has broken constant folding of these casts.

I assume it would be possible to disable this transform if the operand is constant to restore the const folding, or I could try figuring out how to juggle it with DecomposeLongs.  Unless there's another alternative?

Reply:
It seems wrong to add this optimization as part of decompose long. But as you point out, it seems like there is no better place for it right now. Perhaps the JIT needs a post-rationalize, pre-decompose/lower optimization phase -- a "late morph" phase. Presumably there are some optimizations done currently in lower that could be moved there.